gadget/layout.go

```diff
@@ -189,6 +192,69 @@ func LayoutVolumePartially(volume *Volume, constraints LayoutConstraints) (*Part
 	return vol, nil
 }
 
+func resolveOne(gadgetRootDir, kernelRootDir string, kernelInfo *KernelInfo, pathOrRef string) (string, error) {
+	// content may refer to "$kernel:<name>/<content>"
+	if strings.HasPrefix(pathOrRef, "$kernel:") {
+		kernelRef := strings.SplitN(pathOrRef, ":", 2)[1]
+		l := strings.SplitN(kernelRef, "/", 2)
+		wantedAsset := l[0]
+		wantedContent := l[1]
+		kernelAsset, ok := kernelInfo.Assets[wantedAsset]
+		if !ok {
+			return "", fmt.Errorf("cannot find %q in kernel info from %q", wantedAsset, kernelRootDir)
+		}
+		if !strutil.ListContains(kernelAsset.Content, wantedContent) {
+			return "", fmt.Errorf("cannot find wanted kernel content %q in %q", wantedContent, kernelRootDir)
+		}
+		return filepath.Join(kernelRootDir, wantedContent), nil
+	}
+
+	return filepath.Join(gadgetRootDir, pathOrRef), nil
+}
+
+// ResolveContentPaths resolves any "$kernel:" refs in the gadget
+// content and populates VolumeContent.resolvedSource with absolute
+// paths.
+//
+// XXX: move into LayoutVolume(), operator on *Volume and make private?
+func ResolveContentPaths(gadgetRootDir, kernelRootDir string, lv *LaidOutVolume) error {
+	var kernelInfo *KernelInfo = &KernelInfo{}
+	if kernelRootDir != "" {
+		ki, err := ReadKernelInfo(kernelRootDir)
+		if err != nil {
+			return err
+		}
+		kernelInfo = ki
+	}
+
+	for i := range lv.Volume.Structure {
+		if err := resolveContentPathsForStructure(gadgetRootDir, kernelRootDir, kernelInfo, &lv.Volume.Structure[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func resolveContentPathsForStructure(gadgetRootDir, kernelRootDir string, kernelInfo *KernelInfo, ps *VolumeStructure) error {
+	for i := range ps.Content {
+		source := ps.Content[i].Source
+		if source != "" {
+			newSource, err := resolveOne(gadgetRootDir, kernelRootDir, kernelInfo, source)
+			if err != nil {
+				return err
+			}
+			if strings.HasSuffix(source, "/") {
```

Comment:
Implies you could specify `$kernel:asset/content/` where content is a directory, can kernel.yaml specify a directory in the content too? eg. like so:

```
assets:
    edition: <num>
    content:
        - <dir>/ <-- directory?
```

Reply:
I think those have to match, i.e. if the gadget asks for a dir the kernel needs to provide a dir. Which the current implementation should do. Or am I missing something?

Reply:
What do we do when there is an existing directory (e.g. overlays/) and there are updates. Do we replace the directory or replace the files that we know about? What happens if a file is renamed across revisions? Will we keep accumulating old files this way?

Reply:
This does not change anything, it will do what we currently do. But yeah, there might be issues here, I'm happy to add a TODO